cpp/src/io/parquet/page_data.cu

```diff
@@ -99,6 +102,59 @@ struct page_state_s {
   PageNestingDecodeInfo* nesting_info;
 };
 
+struct delta_binary_state_s {
+  uint8_t const* block_start;  // start of data, but updated as data is read
+  uint8_t const* block_end;    // end of data
+  uint32_t block_size;         // usually 128, must be multiple of 128
+  uint32_t mini_block_count;   // usually 4, chosen such that block_size/mini_block_count is a
+                               // multiple of 32
+  uint32_t values_per_mb;      // block_size / mini_block_count, must be multiple of 32
+  uint32_t value_count;        // total values encoded in the block
+  int64_t last_value;          // last value decoded, initialized to first_value from header
+                               // should be int128_t, but in all likelihood we'll never see
+                               // an int96 with this encoding
+  uint32_t current_value_idx;  // current value index, initialized to 0 at start of block
+
+  int64_t cur_min_delta;         // min delta for the block
+  uint32_t cur_mb;               // index of the current mini-block within the block
+  uint8_t const* cur_mb_start;   // pointer to the start of the current mini-block data
+  uint8_t const* cur_bitwidths;  // pointer to the bitwidth array in the block
+
+  uint64_t value[non_zero_buffer_size];  // circular buffer of delta values
+};
+
+struct delta_byte_array_state_s {
+  delta_binary_state_s prefixes;          // state of decoder for prefix lengths
+  delta_binary_state_s suffixes;          // state of decoder for suffix lengths
+  uint64_t offset[non_zero_buffer_size];  // circular buffer for string output offsets
```

Comment:
`value` is unbounded.  I'm still wondering what happens where you have adjacent original values of LONG_MAX and LONG_MIN, where the delta will be huge. If `min_delta` is then negative, there would be overflow.  I have to read the spec closer to see how that would be handled.

`offset` is an offset into a per-page buffer used to store the fully formed intermediate strings. This data will then be copied into the column data later on when `make_strings_column` is called. So I guess this could be `size_type`, if that's where your question was leading.